lib/srv/alpnproxy/conn_test.go

```diff
@@ -157,8 +157,21 @@ func TestPingConnection(t *testing.T) {
 				buf := make([]byte, readSize)
 				for {
 					n, err := r.Read(buf)
-					if err != nil && !errors.Is(err, io.EOF) {
-						require.Fail(t, "Failed to read from connection: %v", err)
+					if err != nil {
+						switch {
+						// Since we're partially reading the message, the last
+						// read will return an EOF. In this case, do nothing
+						// and send the remaining bytes.
+						case errors.Is(err, io.EOF):
+						// The connection will be closed only if the test is
+						// completed. The read result will be empty, so return
+						// the function to complete the goroutine.
+						case errors.Is(err, io.ErrClosedPipe):
+							return
+						// Any other error should fail the test.
+						default:
+							require.Fail(t, "Failed to read from connection: %v", err)
```

Comment:
You can't call `require.Fail` in a goroutine. The proper way to do this would be to send an error to a channel that you read from in the main test goroutine, and fail the test from there.